### ISSUE TITLE
Clarify intended semantics of `document.hasStorageAccess`

### ISSUE THREAD

**Issue by cfredric** [privacycg/storage-access#171]
What are the *intended* semantics / meaning for `document.hasStorageAccess`?

The spec currently says: "This specification defines a method to query whether or not a [Document](https://dom.spec.whatwg.org/#document) currently has access to its [unpartitioned data](https://privacycg.github.io/storage-access/#unpartitioned-data) ([hasStorageAccess()](https://privacycg.github.io/storage-access/#dom-document-hasstorageaccess)) ..."

However, to my knowledge the spec has never included a check to see if unpartitioned cookies *really are* accessible. This means that the spec makes some assumptions that aren't necessarily true, which can invalidate the aforementioned description.
* E.g., the user agent might allow the user to block all cookies (or block cookies on specific sites). Since there's no step in `document.hasStorageAccess` to check whether such a user setting is blocking cookies (even if the script successfully called `document.requestStorageAccess`, or if the script is in a top-level document), `document.hasStorageAccess` might return `true`, even though cookie access is still blocked in that context.
* Inversely, the user agent might allow the user to unconditionally allow access to unpartitioned cookies on a given site. Since the Storage Access API spec assumes that unpartitioned cookie access is always blocked by default, and does not include a step to test whether a call to `document.requestStorageAccess` is necessary in the first place, `document.hasStorageAccess` might return `false` even though unpartitioned cookies are actually accessible in that context.

The way I see it, there are two ways of looking/thinking about the intent:
1. `document.hasStorageAccess` is just about the frame-local `has storage access` state (modulo the early-returns). It's the way for a frame to tell if it needs to call `document.requestStorageAccess`, even if the storage-access permission has already been granted. But if `document.hasStorageAccess` returns true, that isn't a guarantee that the script is now able to read/write unpartitioned cookies; something else (like a user setting) might still block cookie access.
1. `document.hasStorageAccess` is the single source-of-truth for whether the script can truly access unpartitioned cookies. If it returns true, access is allowed; if it returns false, access is blocked. This is kind of like `navigator.cookieEnabled`, but now it knows about unpartitioned vs partitioned cookies.

I can see arguments for both options:
* Keeping it focused on frame-local state keeps the various APIs orthogonal to each other.
  * E.g., use `navigator.cookieEnabled` to see if you actually have cookie access (though there's weirdness here related to whether it access to partitioned cookies counts, or if it's just access to unpartitioned cookies that counts); `permissions.query` to see if you have the "storage-access" permission; and `document.hasStorageAccess` to see if you've already done the per-frame opt-in.
* Making `document.hasStorageAccess` the "one stop shop" to see if unpartitioned cookies are accessible means the site dev only has to call one API to get the answer.
  * However, a return value of `false` doesn't help devs understand why access is blocked and react accordingly.

I think I lean toward option 1, keeping the APIs more orthogonal, since it seems to give more useful information to site developers. (And it's not too hard to write a function to check whether unpartitioned cookies are accessible, anyway). WDYT?

**Comment by johannhof** [comment]
I don't have a super strong opinion but I agree that we should align on something, so thanks for filing this. Some thoughts in favor of option 2:

What I like about option 2 is that hSA can then properly inform the developer whether calling rSA is necessary. Otherwise, as you mention, the developer would need to check `navigator.cookieEnabled` in addition to hSA to figure that out, which is not only somewhat less convenient (although it is a sync call, which is also nice), but it's also inaccurate in some browsers for compat reasons, and there's a chance that cookieEnabled further evolves to become less reliable in the future, as a binary yes/no cookie signal doesn't reflect the reality of the web. (CC @dcthetall who's looking into this)

`hasStorageAccess` could instead be the signal for developers to figure out if cross-site data access was available (though it would be much nicer if the API was called `Cross-Site Data Access API` or something like that).

**Comment by annevk** [comment]
cc @johnwilander

**Comment by annevk** [comment]
I discussed this with colleagues and it would be reasonable to us if this essentially means "will setting a non-partitioned cookie work".

I think that would require queueing a task for steps that want to call resolve (ideally refactored so we do that only once).

**Comment by bvandersloot-mozilla** [comment]
That sounds like a reasonable meaning of `document.hasStorageAccess` to me as well.
